### ISSUE TITLE
[Tracking] [Performance] Investigate creating our own KeyboardAvoidingView

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#10273]
Coming from https://expensify.slack.com/archives/C01GTK53T8Q/p1659722402266779?thread_ts=1659347778.519049&cid=C01GTK53T8Q

### Outstanding
- https://github.com/Expensify/App/issues/11463

### Fixed
- https://github.com/Expensify/App/issues/11701
- - https://github.com/Expensify/App/issues/11391
- https://github.com/Expensify/App/issues/11094
- https://github.com/Expensify/App/issues/11087
- https://github.com/Expensify/App/issues/11118
- https://github.com/Expensify/App/issues/7572
- https://github.com/Expensify/App/issues/10609
- https://github.com/Expensify/App/issues/10610
- https://github.com/Expensify/App/issues/11722
- https://github.com/Expensify/App/issues/12114
- https://github.com/Expensify/App/issues/11776
- https://github.com/Expensify/App/issues/11057
- https://github.com/Expensify/App/issues/12622
- https://github.com/Expensify/App/issues/12480
- https://github.com/Expensify/App/issues/10905
- https://github.com/Expensify/App/issues/11858
- https://github.com/Expensify/App/issues/11277
- https://github.com/Expensify/App/issues/10670

### Problem
First a few observations:

- Dealing with KeyboardAvoidingViews (different behaviors) is always kind of confusing imo
- Furthermore, we've got our own [wrapper component](https://github.com/Expensify/App/blob/2231d55a6ddf189042c680979e4981c6c00d178b/src/components/KeyboardAvoidingView/index.ios.js) for KeyboardAvoidingView that isn't used anywhere in the codebase
- We've also got this [LoginKeyboardAvoidingView component](https://github.com/Expensify/App/blob/f02b60aa4e5a7c26eb03f20cd989486a10522bd5/src/pages/signin/SignInPageLayout/LoginKeyboardAvoidingView/index.js). The native implementation is just the KeyboardAvoidingView from React Native, and the web/desktop implementation seems kind of funky to me. It just passes a style and contentContainerStyle to the style prop of the React Native implementation, which seems unnecessary since the RN component has both of those props.
- We also use this abandoned KeyboardSpacer component that is not very different from KeyboardAvoidingView  AFAICT
    - *EdIt:* We now have our own customer KeyboardSpacer component

All that is to say, it feels like our usage of keyboard-avoiding stuff is a bit inconsistent/all-over-the-place, and maybe it's a good time to step back and see if we can consolidate some of this and make it a bit more consistent.

Moreover, we've noticed issues such as [this one](https://github.com/Expensify/App/issues/10188) where the keyboard-avoiding stuff isn't necessarily working as expected. There have been other similar bugs in the past. For example, we have some screens where we use timeouts to wait for navigation transitions to finish before opening the keyboard, etc... Furthermore, many of these bugs have been sort of inconsistent (maybe suggesting that they are caused by race conditions). Lastly, I've noticed that the keyboard is just really sluggish and slow to open on iOS.

### Solution
Now this part is more of a hunch, but I have a theory that many of these bugs are caused because Reanimated (the animation library used by react-navigation) and the React Native KeyboardAvoidingView/KeyboardSpacer components aren't playing nice together. Why? Because Reanimated runs animations synchronously on the UI thread, while the KeyboardAvoidingView components will update asynchronously between the UI thread and the layout/render thread. This makes it hard to keep them in sync / behaving as expected.

I did a bit of research and noticed that Reanimated 3 is [set to introduce a `useAnimatedKeyboard` hook](https://docs.swmansion.com/react-native-reanimated/docs/next/api/hooks/useAnimatedKeyboard/). This would allow us to create a simple KeyboardAvoidingView of our own that is animated synchronously on the UI thread. I think this would be better for performance (which I have noticed is lackluster on iOS staging+production), and hopefully make opening/closing the keyboard much snappier. Furthermore, since (all?) our animations would be running in Reanimated, I have a hunch that many of these inconsistent bugs might go away.

So my proposal (or at minimum, an idea that I think is worth looking into) is that we:

- upgrade to Reanimated 3 (after [this PR](https://github.com/Expensify/App/pull/9841) is merged)
- create our own KeyboardAvoidingView component using the useAnimatedKeyboard hook from Reanimated
- Use our component everywhere. No more use of the RN built-in KeyboardAvoidingView or the KeyboardSpacer library.

**Comment by roryabraham** [comment]
HOLD on https://github.com/Expensify/App/pull/9841

**Comment by roryabraham** [comment]
Taking this off HOLD

**Comment by roryabraham** [comment]
not prioritized yet

**Comment by roryabraham** [comment]
cc @szymon20000

**Comment by marcaaron** [comment]
Thanks Rory, the solution part of this is really fascinating and it sounds like maybe there are some big performance gains hiding away in Reanimated 3. 

I am having a little trouble with the "problem" section of this and think it needs refining. Or at a minimum break concerns into separate issues to tackle one at a time. I'm admittedly having trouble believing that we aren't just holding something wrong (as that's usually the case).

>Dealing with KeyboardAvoidingViews (different behaviors) is always kind of confusing imo

Can we unlock this somehow? There must be some basic cases where the behaviors should be used and it seems possible to articulate what they do and develop some best practices and guidance?

>Furthermore, we've got our own [wrapper component](https://github.com/Expensify/App/blob/2231d55a6ddf189042c680979e4981c6c00d178b/src/components/KeyboardAvoidingView/index.ios.js) for KeyboardAvoidingView that isn't used anywhere in the codebase

Could we figure out if we can delete this and replace it with a regular `KeyboardAvoidingView` ? That seems like a good cleanup task.

>We've also got this [LoginKeyboardAvoidingView component](https://github.com/Expensify/App/blob/f02b60aa4e5a7c26eb03f20cd989486a10522bd5/src/pages/signin/SignInPageLayout/LoginKeyboardAvoidingView/index.js). The native implementation is just the KeyboardAvoidingView from React Native, and the web/desktop implementation seems kind of funky to me. It just passes a style and contentContainerStyle to the style prop of the React Native implementation, which seems unnecessary since the RN component has both of those props.

This too?

>We also use this abandoned KeyboardSpacer component that is not very different from KeyboardAvoidingView AFAICT
EdIt: We now have our own customer KeyboardSpacer component

And this?

>maybe it's a good time to step back and see if we can consolidate some of this and make it a bit more consistent

Agree, but what does it have to do with rolling our own `KeyboardAvoidingView`? Is it possible to articulate what it is exactly about the component that isn't working for us? I think a lot of the changes you are mentioning might have been implemented when many of us were very naively approaching `react-native` and trying to get things done quickly. 

>For example, we have some screens where we use timeouts to wait for navigation transitions to finish before opening the keyboard

What's an example of this? Is it something we have or can create issues to address/improve?

>Furthermore, many of these bugs have been sort of inconsistent (maybe suggesting that they are caused by race conditions)

Could use an example here as well.

>I've noticed that the keyboard is just really sluggish and slow to open on iOS.

And it's because of the `KeyboardAvoidingView` ? Do we have an open issue for this?

----

One thing learned from today is that barely anyone has a proper understanding of how this component works. 😂 So I'd love to keep exploring `KeyboardAvoidingView` and see if it is the cause of these woes or if we can make it work for us or not.

**Comment by roryabraham** [comment]
No update here yet, but looking forward to working on this. Fortunately `useAnimatedKeyboard` was [released](https://github.com/software-mansion/react-native-reanimated/releases) in `2.10` so upgrading to a `3.0` beta release is no longer a prerequisite of doing this

**Comment by tgolen** [comment]
@roryabraham I am going to grab this from you and start digging into this.

**Comment by roryabraham** [comment]
Cool, thanks! Just FYI, we have already upgraded to Reanimated 2.10, which includes the `useAnimatedKeyboard` hook.

**Comment by JmillsExpensify** [comment]
@tgolen We have a bunch of issues we are holding on for this one. I'm thinking of linking them all in the OP, for clarify on the scope of what keyboard issues/bugs we're talking about. Is that ok with you?

**Comment by tgolen** [comment]
I think that's OK, yep! That'll help me out a lot.

**Comment by JmillsExpensify** [comment]
Sweet, I'm on it!

**Comment by tgolen** [comment]
I'm making pretty good progress on this. I've mostly spent my time ensuring that I can test everything and get a handle on what's broken and what's not.

I've spent a fair amount of time just refactoring our code to use the React-Native `KeyboardAvoidingView` everywhere consistently. Once it's used everywhere, I will go through all the open issues and see what's reproducible and what isn't. Then, see about digging into the bugs and see if they can be fixed with the RN component, or if we feel like building our own basic component (based off the `useAnimatedKeyboard` hook) is a route to go.

**Comment by JmillsExpensify** [comment]
Circling back to check how things have been going since your last update?

**Comment by tgolen** [comment]
Ah, yes. I had been waiting for some changes from Jules to be merged in another related PR https://github.com/Expensify/App/pull/10648 which prevents me from doing any definitive testing on this feature. This is currently on HOLD until that PR is merged.

**Comment by JmillsExpensify** [comment]
What are the next steps on this issue? I peek at the referenced PR and that's still in the works. Does our plan change at all in light of @Julesssss comment above?

**Comment by Julesssss** [comment]
Cool, that makes sense to me, but FYI I'm unlikely to be able to prioritize my PR anytime soon. It might be worth prioritizing this PR over mine, on the assumption that my change could be on hold for months.

**Comment by tgolen** [comment]
Ah, that's really good to know. But, why months? I can move forward with mine if that's the case

**Comment by Julesssss** [comment]
Purely because I'm leading two high-priority roadmap-blocking projects at the moment: [WhatsApp Quality] Notifications Rework and then [N7: Free Plan] Manual Requests (in addition to finishing up the Manual Requests API Refactors).

**Comment by JmillsExpensify** [comment]
Given that Manual Request is on hold with the exception of the API Refactors, and then related, this PR helps unblock a significant priority within WhatsApp Quality, why not prioritize this PR now? My concern with holding on this for months is that we have a ton of keyboard-related issues that you can help unblock. Yes, notifications are valuable too but I'd rather see us prioritize unblocking things that are currently in motion. Well that or I would consider that you unassign and we can help highlight for someone to pick up. My two cents!

**Comment by Julesssss** [comment]
That's fair. I also think we should continue to prioritize this 'Create our own KeyboardAvoidingView' issue -- without holding for my native fix PR. 

While I was originally confident in my native fix, with the latest PR review I think I was mistaken in believing it fully resolved the original issue. 😞 I would be more than happy to hand it off to someone else, but I worry it is a potential dead end and for that reason don't think it's worth prioritizing, personally.

**Comment by Julesssss** [comment]
> I prefer having Jules's PR merged first because it is a change to the underlying foundation and my changes are higher up in the view hierarchy, so I think Jules's changes will have a big impact on consistently trying to reproduce all these issues.

I think we should rethink this, with the new assumption that the native fix is not going to work. I'm happy to help advise any engineer that picks up the native issue, but my confidence in this being a viable solution is now low.

**Comment by tgolen** [comment]
OK, after discussing it more with Jules, I'm going to take this off hold and bump it up to daily priority.

**Comment by JmillsExpensify** [comment]
Awesome, great discussion!

**Comment by tgolen** [comment]
OK, I started testing all the GH issues today with my code changes. I made it through the first 5 issues. I found that 2 of those issues are fixed, 2 are not fixed, and one is expected behavior.

I've got 6 more issues to go through 👍 so plugging away.

Ultimately, my changes do put us in a better place so far, so that's good.

**Comment by tgolen** [comment]
I finished going through all the issues this morning. In the end, my PR fixes 4 of the issues, so it's a step in the right direction. My plan is to go ahead and get my PR merged to fix those 4 issues, then remove the rest of them off HOLD so they can continue to be worked on, using the new cleaned-up code from my PR.

**Comment by JmillsExpensify** [comment]
Ok that sounds great! Happy to handle the remaining issues when your PR is merged.

**Comment by tgolen** [comment]
Reviews are happening! There was just a minor bug introduced with some styles on the login page, but it's in final review now.

**Comment by tgolen** [comment]
I haven't had the chance to progress on the PR this week due to allocations.

**Comment by JmillsExpensify** [comment]
@tgolen Now that allocations are done, are you prioritizing this one?

**Comment by tgolen** [comment]
Apologies for being gone last week without notice. This is near the top of my priority list yes. The PR is still being reviewed and tested. A couple of issues were found on the iPad still and one small issue on Android. We're deciding how best to proceed with those. My hope is to have it merged today or tomorrow.

**Comment by tgolen** [comment]
I'm pushing every day to get my PR merged! It feels like we're close (within a day), but it's dragging out with the testing/reviewing.

**Comment by tgolen** [comment]
OK, my PR has been merged, so I am going to remove the HOLD on most of these issues. They can move forward with getting proposals for them and I would like to be included on most of them to ensure the proposals are not hacks or workarounds.

**Comment by JmillsExpensify** [comment]
Going to assign myself as well given that this is a major WAQ initiative and we have 11 issues that are being opened up for solutions now.

**Comment by JmillsExpensify** [comment]
Went through all linked issue and updated / pushed on next steps.

**Comment by JmillsExpensify** [comment]
Went through several of the linked issue and was able to close them by finding either dupes or not being able to reproduce the issue anymore.

**Comment by JmillsExpensify** [comment]
Did a sweep of the issues just now. Most are stalled at the proposal stage. I've gone through and raised prices on the majority of them.

**Comment by flodnv** [comment]
Should https://github.com/Expensify/App/issues/11463 be linked in this GH issues? I think so but not 100% sure

**Comment by tgolen** [comment]
Yeah, it looks like it should be linked to this issue. I'll add it

**Comment by Julesssss** [comment]
The vast majority are still awaiting proposals. One issue was resolved, and another was unreproducible so I asked Applause to retest.

**Comment by JmillsExpensify** [comment]
Vast majority still awaiting proposals. I floated this one with Margelo and also highlighted in the weekly update.

**Comment by hannojg** [comment]
Hey guys, Hanno from Margelo here 👋 
Is someone in your team already working on a custom KAV or is this something that we could start on working? 😊

**Comment by tgolen** [comment]
At this point, I still haven't determined if creating a custom KAV is necessary. I think we're getting pretty good mileage out of the official KAV component from RN so I wouldn't put any time into doing anything custom yet.

**Comment by hannojg** [comment]
Alright, we still have this on our todo to look into all these issues and try to come up with a solution that will help with all/most of them. Expect more updates next week

**Comment by JmillsExpensify** [comment]
Thanks @hannojg! Agree with that plan.

**Comment by chrispader** [comment]
I will further investigate this issue and coordinate with @hannojg first 👍

**Comment by JmillsExpensify** [comment]
We're all assigned out and working through individual issues.

**Comment by tgolen** [comment]
Still plugging away on this!

**Comment by chrispader** [comment]
Sorry, i searched a meaningful translation of *plugging away* and *working away*, but couldn't exactly figure out what you mean 😅

Is this issue something we should try to fix right now or should we wait until all of the affected issues are solved?

**Comment by tgolen** [comment]
I think all of the affected issues are assigned out and being worked on by someone already so there isn't anything to do here currently except wait, @chrispader. 

Another way of saying "plugging away" or "working away" would be: This is being worked on and progress is being made on it.

**Comment by chrispader** [comment]
Ok, thanks! @tgolen

**Comment by JmillsExpensify** [comment]
Ah sorry about that @chrispader! Regardless, agree with Tim. I think we're largely good on the related KAV issues at this point. We'll let you know if anything else comes up. 😄

**Comment by JmillsExpensify** [comment]
Circling back to our outstanding issues above, we continue to make progress. Only a handful of issues left.

**Comment by JmillsExpensify** [comment]
Great process all around for this initiative. We're down to three issue in progress, and one issue that is held on linked issues. While I'm not confident that we'll close this out before the end of the year, we should be able to close this shortly thereafter.

**Comment by tgolen** [comment]
There are 4 remaining issues. One is in review, one is on hold. The other two are being worked on by Margelo it looks like.

**Comment by tgolen** [comment]
No change since yesterday.

**Comment by JmillsExpensify** [comment]
Comment from last week still holds. 4 linked issues remain outstanding, though at least one should be closed out soon and then 2 more are getting close. The one held issue will likely remain that way for a while given that it's blocked on upstream changes, which are dependent on third party repos treating this as a priority.

**Comment by JmillsExpensify** [comment]
Same as above.

**Comment by JmillsExpensify** [comment]
We have two outstanding unheld issues above, so we continue to make progress. Both of those issues are in various stages of review, which is great. They aren't holding up WAQ! Overall, I expect slow progress this week given that most of the issue assignees are out on vacation until the new year.

**Comment by JmillsExpensify** [comment]
Still working through the two linked issues in the OP.

**Comment by JmillsExpensify** [comment]
Quick update. One of the remaining issues should be wrapped up and close in the coming days, and likely similar for the other one. At that point we're only blocked on the one remaining issue still on hold.

**Comment by JmillsExpensify** [comment]
One more issue remains that isn't on hold. Related, that held issue should be coming off hold this week.

**Comment by JmillsExpensify** [comment]
PR for the last non-held issue is on staging and should hit production tomorrow!

**Comment by JmillsExpensify** [comment]
Working through a regression on the last outstanding issue.

**Comment by JmillsExpensify** [comment]
Last issue is getting close! Still working through that regression, though it's likely to be merged this week.

**Comment by janicduplessis** [comment]
What is the status on this? I ended up investigating using `useAnimatedKeyboard` while working on enabling edge to edge display on android. However it seems to cause an issue where the status bar is no longer transparent as soon as we use the hook. So it would probably require a fix in reanimated. Except for that it seems to work really well on Android and adds a nice animation so the content moves with the keyboard opening.

I will leave the code here for later use if we can have a fix in reanimated.

```js
/*
 * The KeyboardAvoidingView is only used on ios and android
 */
import React from 'react';
import Animated, {
    useAnimatedKeyboard,
    useAnimatedStyle,
} from 'react-native-reanimated';

const KeyboardAvoidingView = (props) => {
    const keyboard = useAnimatedKeyboard({isStatusBarTranslucentAndroid: true});
    const animatedStyle = useAnimatedStyle(() => ({
        paddingBottom: keyboard.height.value,
    }));
    return (
        // eslint-disable-next-line react/jsx-props-no-spreading, react/prop-types
        <Animated.View {...props} style={[props.style, animatedStyle]} />
    );
};

KeyboardAvoidingView.displayName = 'KeyboardAvoidingView';

export default KeyboardAvoidingView;
```

**Comment by JmillsExpensify** [comment]
In terms of this issue, all linked issues above are merged and nothing is being actively worked on for `KeyboardAvoidingView`. We're essentially just working through a regression/payment step in the one held issue mentioned above.

**Comment by roryabraham** [comment]
@tomekzaw Want to take a look at @janicduplessis's bug report here?

**Comment by tomekzaw** [comment]
@roryabraham Sure, I will be more than happy to help here.

You're right, https://github.com/software-mansion/react-native-reanimated/pull/3958 is already merged and released so we'll just need to upgrade to 3.0.2 first.

As for the `useAnimatedKeyboard` hook, we're still investigating two minor issues:
1. It looks like there is 1 or 2 frames of delay between keyboard animation start and shared value update.
2. After some configuration of opening/closing the keyboard, the shared value gets updated only 3-4 times during the animation (see https://github.com/software-mansion/react-native-reanimated/issues/4194)

**Comment by terrysahaidak** [comment]
@tomekzaw I have KeyboardAvoidingView for iOS implemented using `useAnimatedKeyboard` in my PR https://github.com/Expensify/App/pull/16356 since my task relies heavily on the keyboard being animated in the same frame as everything else, I wanna animated based on its height.

**Comment by MelvinBot** [comment]
📣 @terrysahaidak! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by JmillsExpensify** [comment]
As for the this tracking issue, I kind of think we should close this issue since all the related issues are now done. Maybe we should create a separate issue for upgrading Reanimated?

**Comment by terrysahaidak** [comment]
As part of this issue and my #16356, we (cc @roryabraham ) decided to contribute the implementation of KeyboardAvoidingView I did as part of the PR to Reanimated. This way they will be able to track and fix all the bugs related to `useAnimatedKeyboard` hook. I'll add PR later today.

Another thing I did - reported a bug that stops us from upgrading Reanimated version which fixes some other bugs:
https://github.com/software-mansion/react-native-reanimated/issues/4324

**Comment by tomekzaw** [comment]
> Another thing I did - reported a bug that stops us from upgrading Reanimated version which fixes some other bugs:
https://github.com/software-mansion/react-native-reanimated/issues/4324

Fixed with https://github.com/software-mansion/react-native-reanimated/pull/4329

**Comment by JmillsExpensify** [comment]
Someone please re-open and correct me if I'm wrong!

**Comment by terrysahaidak** [comment]
Here is the PR that implements KeyboardAvoidingView in Reanimated:
https://github.com/software-mansion/react-native-reanimated/pull/4339

**Comment by terrysahaidak** [comment]
The implementation of KeyboardAvoidingView is pretty much ready, finishing the example used for testing, it mimics our usage I'm working at #16356 .

Also, we're currently testing this KeyboardAvoidingView implementation on Android on another project.
